text/0000-multi-crate-impls.md

```diff
@@ -0,0 +1,147 @@
+- Feature Name: coupled_workspaces
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is a way to more tightly couple crates within a cargo workspace, allowing
+inherent impls and blanket trait impls across multiple crates in the same
+project. Its primary use case is for the standard library, but the feature is
+extended so that it can be used by the entire crate ecosystem.
+
+# Motivation
+[motivation]: #motivation
+
+Many projects have grown large enough to split their contents across multiple
+crates, so that users of these crates can decide which parts they want to use
+and which ones they don't. Notable examples on crates.io include the `num` and
+`unic` crates, but the standard library is also split into several crates.
+
+Unfortunately, this brings up a problem: inherent methods must be contained
+within the crate that defines a type. This causes a lot of friction when
+multiple crates build upon the functionality of each other, like the mess
+between `std_unicode`, `alloc`, and `core`, which all provide varying levels of
+functionality for `str` and `char`.
+
+While a solution for the standard library is necessary, the ecosystem could
+benefit from this functionality as well, and extending the concept of cargo
+workspaces to include this use case seems natural.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC won't re-explain Cargo's workspace functionality, which is already
+implemented and documented. However, a few particularly notable changes arise:
+
+1. `pub(crate)` definitions become public to all crates in a workspace.
+2. Traits and types are, for all intents and purposes, considered "in the same
+   crate" as all crates in their workspace. This means that inherent impls are
+   allowed across crates, and a trait can be implemented for a type as long as
+   the type or the trait are in the same workspace, not just the same crate.
+3. Workspaces which use virtual manifests must opt into this feature by
+   specifying `workspace.root` in `Cargo.toml`.
+4. Workspaces can opt out of this with `workspace.coupled = false` in
```

Comment:
I think default behavior should be considered. For instance build scripts named build.rs don't need to be named in Cargo.toml anymore.

I think we should err on the side of opt in rather than opt out, but I really do like this in general and would clean up some of my codebases if this was available.

Reply:
This is an artefact of the fact that I wasn't sure what the default behaviour should be when writing this; in the guide-level explanation I don't list the default but in the reference-level I do.

I'll fix this later today; I agree that it should be opt-in, as it prevents the workspace from allowing multiple versions of crates.